### PR TITLE
[Memory 1/N] Create slot_map data structure

### PR DESCRIPTION
**Pull request by Phantomical**
See the doc comment for slotmap for an explanation of what this does and the comment within for an explanation of how it's implemented. I've included some tests that should cover most of the general use, although they might not cover all of the edge cases.

This is needed as part of the memory system.

### CODE REVIEW

File: include/caffeine/ADT/SlotMap.h [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+#include <deque>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+namespace caffeine {
+
+namespace detail {
+  template <typename F>
+  class drop_guard : F {
```
Comment (mishazharov) [inline include/caffeine/ADT/SlotMap.h:15]: What's this class for? Could use more documentation

File: include/caffeine/ADT/SlotMap.h [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+   */
+  template <typename T>
+  constexpr T replace(T& loc, const T& value) {
+    T old = std::move(loc);
+    loc = value;
+    return old;
+  }
+
+  template <typename T>
+  using remove_cvref_t = std::remove_cv_t<std::remove_reference_t<T>>;
```
Comment (mishazharov) [inline include/caffeine/ADT/SlotMap.h:45]: What's this responsible for? Can't find any other references to it

**Review by mishazharov — APPROVED** [review]
LGTM otherwise

File: include/caffeine/ADT/SlotMap.h (hunk 2 above)
Comment (Phantomical) [reply include/caffeine/ADT/SlotMap.h]: This was used at some point during development but I guess that code got rewritten so it's just dead code then.
